metrics_v2/publisher/event_publisher.py

```diff
@@ -0,0 +1,200 @@
+flags.DEFINE_string('pubsub_topic', None, 'PubSub topic to publish to.')
+flags.DEFINE_string('model_output_bucket', None, 'GCS location of model outputs.')
+
+flags.DEFINE_string('project', None, 'GCP Project ID of cluster and PubSub topic.')
+flags.DEFINE_string('cluster_name', None, 'Name of current cluster.')
+flags.DEFINE_string('cluster_location', None, 'Location of current cluster.')
+flags.DEFINE_string('namespace', 'default', 'Namespace to watch for Jobs.')
+
+
+def _resource_version_path():
+  if FLAGS.save_dir:
+    return os.path.join(FLAGS.save_dir, 'resource_version.txt')
+  else:
+    return 'resource_version.txt'
+
+def _save_resource_version(resource_version: int):
+  logging.debug('Saving resource_version %d', resource_version)
+
+  with open(_resource_version_path(), 'w') as f:
+    f.write(str(resource_version))
+
+def _load_resource_version() -> typing.Optional[int]:
+  try:
+    with open(_resource_version_path(), 'r') as f:
+      return int(f.read())
+  except FileNotFoundError as e:
+    logging.warning('Resource version file not found.', exc_info=e)
+    return None
+
+def _get_metadata(attribute):
+  resp = requests.get(
+      f'http://metadata.google.internal/computeMetadata/v1/instance/attributes/{attribute}',
+      headers={'Metadata-Flavor': 'Google'})
+  resp.raise_for_status()
+
+  return resp.text
+
+def create_test_completed_event(
+    job: kubernetes.client.V1Job,
+    model_output_bucket: str,
+    cluster_name: str,
+    cluster_location: str,
+    project:str
+) -> metrics_pb2.TestCompletedEvent:
+  """Returns a TestCompletedEvent to publish to PubSub.
+
+  Args:
+    job: A Kubernetes Job resource.
+    model_output_bucket: Path to GCS bucket with model outputs.
+    cluster_name: Name of the current Kubernetes cluster.
+    cluster_location: Location (region or zone) of the current Kubernetes cluster.
+    project: The project ID of the current project.
+  
+  Returns:
+    A TestCompletedEvent with the information from job.
+  """
+  if len(job.status.conditions) == 1:
+    condition = job.status.conditions[0]
+  # job.status.conditions _usually_ has length 1, but it can have both passing and failing conditions.
+  # Give precedence to failing conditions.
+  elif len(job.status.conditions) == 0:
+    logging.error('Job %s has no conditions.', job.metadata.name)
+    return
+  else:
+    condition = next((c for c in job.status.conditions if c.type == 'Failed'), None)
+
+  if condition.reason == 'DeadlineExceeded':
+    job_status = metrics_pb2.TestCompletedEvent.TIMEOUT
+  elif condition.reason == 'BackoffLimitExceeded':
+    job_status = metrics_pb2.TestCompletedEvent.FAILED
+  elif condition.type == 'Complete':
+    job_status = metrics_pb2.TestCompletedEvent.COMPLETED
+  else:
+    logging.error('Unknown condition for Job %s: %s', job.metadata.name, str(condition))
+    return
+
+  annotations = job.metadata.annotations or {}
+  gcs_subdir = annotations.get('ml-testing-accelerators/gcs-subdir', '')
```

Comment:
possible to move this to a named constant? Might make it easier to find if someone is updating the repo to run their own tests.

Ignore if there is no need to update this string. It looks like this is just a generic key and doesn't need to match a repo name or project name

Reply:
The prefix will be the same for any user/repo/project. I added it to keep our annotations from colliding with other projects'